syntax = "proto3";

package TW.Tezos.Proto;
option java_package = "wallet.core.jni.proto";

// Input data necessary to create a signed Tezos transaction.
// Next field: 3
message SigningInput {
  OperationList operation_list = 1;
  bytes private_key = 2;
}

// Transaction signing output.
// Next field: 2
message SigningOutput {
  bytes signed_bytes = 1;
}

// A list of operations and a branch.
// Next field: 3
message OperationList {
  string branch = 1;
  repeated Operation operations = 2;
}

// An operation that can be applied to the Tezos blockchain.
// Next field: 12
message Operation {
  int64 counter = 1;
  string source = 2;
  int64 fee = 3;
  int64 gas_limit = 4;
  int64 storage_limit = 5;

  enum OperationKind {
    // Note: Proto3 semantics require a zero value.
    ENDORSEMENT = 0;

    REVEAL = 7;
    TRANSACTION = 8;
    ORIGINATION = 9;
    DELEGATION = 10;
  }
  OperationKind kind = 7;

  // Operation specific data depending on the type of the operation.
  oneof operation_data {
    RevealOperationData reveal_operation_data = 8;
    TransactionOperationData transaction_operation_data = 9;
    OriginationOperationData origination_operation_data = 10;
    DelegationOperationData delegation_operation_data = 11;
  }
}

// Transaction operation specific data.
// Next field: 3
message TransactionOperationData {
  string destination = 1;
  int64 amount = 2;
}

// Reveal operation specific data.
// Next field: 2
message RevealOperationData {
  bytes public_key = 1;
}

// Origination operation specific data.
// Next field: 3
message OriginationOperationData {
  string manager_pubkey = 1;
  int64 balance = 2;
}

// Delegation operation specific data.
// Next field: 2
message DelegationOperationData {
  string delegate = 1;
}